### ISSUE TITLE
Execution of test methods in different classes is interleaved when dependsOnMethod is used

### ISSUE THREAD

**Issue by sleske** [cbeust/testng#106]
I am seeing an odd order of test executions in TestNG.

I created a simple project, with this testng.xml:

```
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
<suite name="Simple suite">
    <test name="Simple tests">
        <packages>
            <package name="testngtests"></package>
        </packages>
    </test>
</suite>
```

The package testngtests contains classes MyTest1, MyTest2 with several test methods, all of which are empty:

```
@Test
public void testEins(){

}
```

etc.

Two of the tests use "dependsOnMethods":

```
@Test(dependsOnMethods ="testEins")
public void testZwei(){}

@Test(dependsOnMethods ="testZwei")
public void testDrei(){}
```

If I run the suite using TestNG V 5.10-jdk15 or V 6.0.1, they always run in the same order:

```
testngtests.MyTest1.testEins()
testngtests.MyTest1.testZwei()
testngtests.MyTest1.testDrei() 
testngtests.MyTest2.testEinsA2()
testngtests.MyTest2.testDeins2()
testngtests.MyTest2.testDreiB2() 
testngtests.MyTest2.testZweiZ2()
```

as expected.

However, with TestNG 6.2.1, the order is:

```
MyTest1.testEins()
MyTest2.testDeins2()
MyTest2.testDreiB2()
MyTest2.testEinsA2()
MyTest2.testZweiZ2()
MyTest1.testZwei()
MyTest1.testDrei()
```

Note that now the tests from different classes are interleaved.

I expected tests from the same class to always run together (though not always in the same order). From the TestNG manual: "TestNG will try to group test methods by class."

So did some change to TestNG break ordering in 6.2.1?

Or does TestNG simply not guarantee that test will be grouped by class? In that case the docs should be clarified to explain when TestNG will group by class - or not.

Also, in that case it would be nice to have a setting that forces grouping tests by class. Should I file a feature request for that?

**Comment by tautom** [comment]
Hi,

I am seeing the same problem. When using groups, dependsOnMethods in the TestClasses the execution of TestMethods between the TestClasses get interleaved, causing all kinds of issues for our setup. 

Does anyone know which is a good known TestNG VERSION to go back to to avoid this problem??

Thanks a lot for posting about this issue!!

**Comment by tautom** [comment]
The way i got around this is to add @Test(singleThreaded = true) at each TestClass level

**Comment by saberduck** [comment]
I solved this by writing IAnnotationTransformer listener which sets distinct priority to every test class. However I still think that this is testNG issue and by default tests between different classes should not be interleaved, as it goes against intuitiveness

```java
@Override
  public void transform(ITestAnnotation annotation, Class testClass, Constructor testConstructor, Method testMethod) {
    Class<?> declaringClass = testMethod.getDeclaringClass();
    Integer priority = priorityMap.get(declaringClass);
    if (priority == null) {
      priority = priorityCounter++;
      priorityMap.put(declaringClass, priority);
    }
    log.trace("Setting priority {} for {}", priority, testMethod);
    annotation.setPriority(priority);
  }
```

**Comment by juherr** [comment]
Could someone send a pr with a test that prove the issue? It will be easier to find a fix.

**Comment by rozzamozza** [comment]
Total agree with last post. Interleaved Class tests cause no end of pain.
On a large project tests that run in different order at different times just result in random failures.
I know there is not supposed to be state between tests but here in the real work that does slip in and debugging random failures is not fun.

**Comment by mhiemstrapodiumtv** [comment]
I am using Testng and nearly everything is working really well. There is one serious issue with priority. No matter what I do priority is ignored.

When I say

@Test( groups ="testA", priority=123 )
public void testMethodA() {
  // Do nothing
}

I see this in the output:

my.package.CalendarPage
  @BeforeMethod AbstractPageTest.beforeMethod()[pri:0, instance:my.package.CalendarPage@183b7230]
    @Test CalendarPage.testCalendarItems()[pri:0, instance:my.package.CalendarPage@183b7230]

priority isn't being read. It is set to 0.

So I tried a IMethodInterceptor and a IAnnotationTransformer and in each listener I execute

annotation.setPriority ( 999 );

and still it says

my.package.CalendarPage
  @BeforeMethod AbstractPageTest.beforeMethod()[pri:0, instance:my.package.CalendarPage@183b7230]
    @Test CalendarPage.testCalendarItems()[pri:0, instance:my.package.CalendarPage@183b7230]

My xml contains this:

<suite name="Test suite" parallel="false">

	<listeners>
            <listener class-name="my.package.TestAnnotationTransformerListener" />
  	</listeners>

	<test name="WebsiteTests" verbose="3">
		<parameter name="browser" value="firefox" />

		<groups>
			<run>
				<include name="testA" />
			</run>
		</groups>

		<packages>
			<package name="my.package.*" />
		</packages>
	</test>
</suite>

Either there is a bug in testng with priorities or there is a configuration conflict or something.

**Comment by emmahuang** [comment]
still facing this problem confused!

**Comment by bryannliu** [comment]
Version 6.9.9 still have this issue.

**Comment by juherr** [comment]
@bryannliu Could you share a failing example ?

**Comment by bryannliu** [comment]
Hi @juherr , sorry, My mistake, 6.9.9 works fine. I ran it in IDEA 14 which have testng version 5.8 integrated that cause this issue.
So this issue should be fixed in latest version.

**Comment by abaxanean** [comment]
6.9.10, the issue is present.

**Comment by juherr** [comment]
@abaxanean Could you share an example we can run ?

**Comment by abaxanean** [comment]
I have 5207 test methods, 16 of them marked with dependsOnMethods.
7 of these 16 get constantly interleaved.

**Comment by juherr** [comment]
Ok. But how can I reproduce it on my side ?

**Comment by abaxanean** [comment]
And that's when using testng suite xml. Because without the xml (just let maven run the tests) much more methods are interleaved, regardless of whether they depend on other method(s).

**Comment by abaxanean** [comment]
Another issue, maybe it's related to this one, is that if I have an IMethodInterceptor all methods are passed in parameter, including methods marked with dependsOnMethods. The documentation states : _Only methods that have no dependents and that don't depend on any other test methods will be passed in parameter_.

**Comment by juherr** [comment]
Please, share a sample for the first issue and open a new one for the second.

**Comment by abaxanean** [comment]
https://github.com/abaxanean/testng-dependsOn

**Comment by Dao007forever** [comment]
For @abaxanean example, looks like in TestRunner.java, we do a top sort and run all nodes without dependencies in one run. So if `Test1.test2()` depends on `Test1.test1()`, and `Test3.test3()` has no dependencies, then `Test1.test1()` and `Test3.test3()` will be run first. Is this the expected behavior?

**Comment by abaxanean** [comment]
@Dao007forever 
I think this should definitely not be the expected behavior. This breaks the whole @BeforeClass / @AfterClass concept. There must be just 2 valid orders : 
Test1.test1() Test1.test2() Test3.test3()
OR
Test3.test3() Test1.test1() Test1.test2()

**Comment by juherr** [comment]
@Dao007forever is rigth. `Test1.test1() Test3.test3() Test1.test2()` is legal and expected (expected because of parallel feature).
`@Before/AfterClass` concept only means "BeforeClass" will be before all methods of the class and "AfterClass" will be after.
`Test1.beforeClass() Test1.test1() Test3.beforeClass() Test3.test3() Test3.afterClass() Test1.test2() Test1.afterClass()` is respecting the concept.

**Comment by abaxanean** [comment]
For a parallel execution, sure, interleaving is unavoidable.
But for a sequential one that also doesn't use dependsOnGroups, just dependsOnMethods, isn't it possible to somehow tell testng to avoid interleaving tests from different classes (given that as you say, it is expected to interleave them) ?

**Comment by juherr** [comment]
I can't say it is expected because it is not specified and it can change in the future by design or because of a different jdk target.
The TestNG logic is the following: it creates a dependency graph with the tests. In your case, it is easy:

```
Test1.test1() <- Test1.test2()
Test3.test3()
```

At the first iteration, free nodes are `Test1.test1()` AND `Test3.test3()`. In sequential execution, TestNG will run them in any order. Then, it will run `Test1.test2()`.

It looks you have a dependency between your classes, so you can use another order feature to describe it (ie: priority). And if you don't want to have it in your test code, you can set the values with an `IAnnotationTransformer`.

**Comment by abaxanean** [comment]
Well guys, than just mark the issue as invalid and add a big note to the documentation :

> Even with the default configuration (sequential, no dependencies)TestNG may interleave tests from different test classes, Test1#BeforeClass -> Test2#test2() -> Test1#test1() -> Test1#AfterClass is a perfectly valid execution plan.

**Comment by juherr** [comment]
@abaxanean I've just run your project sample and I see one of your expected behavior: `Test3.test3() Test1.test1() Test1.test2()` 

```
[INFO] --- maven-surefire-plugin:2.19.1:test (default-test) @ dependsOn ---

-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running TestSuite
...
... TestNG 6.9.10 by Cédric Beust (cedric@beust.com)
...

Test3#test3
Test1#test1
Test1#test2
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.258 sec - in TestSuite
```

```
$ mvn --version
Apache Maven 3.3.9 (bb52d8502b132ec0a5a3f4c09453c07478323dc5; 2015-11-10T17:41:47+01:00)
Maven home: /home/juherr/.sdkman/candidates/maven/current
Java version: 1.8.0_111, vendor: Oracle Corporation
Java home: /usr/lib/jvm/java-8-oracle/jre
Default locale: fr_FR, platform encoding: UTF-8
OS name: "linux", version: "4.4.0-51-generic", arch: "amd64", family: "unix"
```

**Comment by abaxanean** [comment]
@juherr 
As you say, this is `expected behavior` not TestNG's. TestNG is also ok with interleaving them.

**Comment by juherr** [comment]
@abaxanean Yes, but as you can see, TestNG is not interleaving classes of your sample.

**Comment by juherr** [comment]
And the `group-by-instances` feature could (should?) allow to force the behavior you expect.

**Comment by abaxanean** [comment]
Adding `group-by-instances` executes just the methods of one test class (for the original project, not this small one)

On my box they do interleave :
```
Running TestSuite
...
... TestNG 6.9.10 by Cédric Beust (cedric@beust.com)
...

Test1#test1
86400000
Test3#test3
Test1#test2 Interleaving detected
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.289 sec - in TestSuite

Results :

Tests run: 3, Failures: 0, Errors: 0, Skipped: 0
```

```
Apache Maven 3.3.9 (bb52d8502b132ec0a5a3f4c09453c07478323dc5; 2015-11-10T08:41:47-08:00)
Maven home: /usr/local/Cellar/maven/3.3.9/libexec
Java version: 1.8.0_112, vendor: Oracle Corporation
Java home: /Library/Java/JavaVirtualMachines/jdk1.8.0_112.jdk/Contents/Home/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "10.11.6", arch: "x86_64", family: "mac"
```

**Comment by juherr** [comment]
Try again with 6.10

**Comment by abaxanean** [comment]
6.10 includes tests from the excluded group.

**Comment by juherr** [comment]
Another issue, another story. Is the interleaved issue fixed on your environment in 6.10 (with or without group-by-instances)?

**Comment by abaxanean** [comment]
With 6.10 the interleaving is still there (the same 7 methods as before)
It also includes tests of the form Test extends A implements B where A is annotation with the excluded group.

**Comment by abaxanean** [comment]
`IAnnotationTransformer` suggested by @saberduck doesn't work because of base classes. The methods of the base class will be executed multiple times but you can specify just 1 priority for them.

**Comment by gdom12345** [comment]
So this is still present in 6.11.  We are not using dependsOnMethod  or any other dependsOn in our tests, but some of our tests have a Priority set.  The only workaround that worked was from @saberduck, but his solution overrides the priority that we set at a test level.  To get around this I re-wrote it to preserve the test level, by concatenating a class priority to the test priority.

Our solution:

package testng_Listeners;
import java.lang.reflect.Constructor;
import java.lang.reflect.Method;
import java.util.HashMap;
import org.testng.IAnnotationTransformer;
import org.testng.Reporter;
import org.testng.annotations.ITestAnnotation;

//Listener to fix TestNG Interleaving issue.  I had to re-write this as the original example I had did not //allow for priority to be manually set on a test level.
public class RePrioritizingListener implements IAnnotationTransformer {

	HashMap<Object, Integer> priorityMap = new HashMap<Object, Integer>();
	Integer class_priorityCounter = 10000;
	// The length of the final priority assigned to each method.
	Integer max_testpriorityLength = 4;

	@Override
	public void transform(ITestAnnotation annotation, Class testClass, Constructor testConstructor, Method testMethod) {

		
		// class of the test method.
		Class<?> declaringClass = testMethod.getDeclaringClass();
		// Current priority of the test assigned at the test method.
		Integer test_priority = annotation.getPriority();
		// Current class priority.
		Integer current_ClassPriority = priorityMap.get(declaringClass);

		if (current_ClassPriority == null) {
			current_ClassPriority = class_priorityCounter++;
			priorityMap.put(declaringClass, current_ClassPriority);
		}

		String concatenatedPriority = test_priority.toString();

		// Adds 0's to start of this number.
		while (concatenatedPriority.length() < max_testpriorityLength) {
			concatenatedPriority = "0" + concatenatedPriority;
		}

		// Concatenates our class counter to the test level priority (example
		// for test with a priority of 1: 1000100001; same test class with a
		// priority of 2: 1000100002; next class with a priority of 1. 1000200001)
		concatenatedPriority = current_ClassPriority.toString() + concatenatedPriority;

		//Sets the new priority to the test method.
		annotation.setPriority(Integer.parseInt(concatenatedPriority));

		String printText = testMethod.getName() + " Priority = " + concatenatedPriority;
		Reporter.log(printText);
		System.out.println(printText);

	}

}

**Comment by michabu** [comment]
After a long time struggling to find a workaround for this behaviour I reverted to using TestNG v6.0.1. At the moment this seems to be a viable solution though I'm afraid I will run into other problems in the long term.

I love that TestNG is able to randomize test-execution-order out of the box. For unit-tests that is. 

Unfortunately I use TestNG not only for unit-tests but also as a driver to execute integration-tests which setup a database and run different steps that depend on each other.
Using IntelliJ IDEA and the `dependsOnMethods`-Annotation that works just fine out of the box.
Using TestNG with gradle on TeamCity on the other hand caused me a lot of headaches and in the meantime also a lot of development hours.

I would like to kindly suggest two alternative approaches to mitigate this @cbeust
1. Prominently and explicitly state in the docs that TestNG is not to be used with stateful integration-tests or introduce a section on how to use it in that case.
2. Offer an easy to configure mode which allows TestNG to be used as an integrationtest driver.

Clearly I would prefer approach 2.) but approach 1.) would at least safe other folks from making the same mistake all over again.

Currently the documentation reads

> TestNG is a testing framework designed to simplify a broad range of testing needs, from unit testing (testing a class in isolation of the others) to integration testing (testing entire systems made of several classes, several packages and even several external frameworks, such as application servers).

I disagree with this statement and would rephrase it as:

> TestNG is a testing framework designed to simplify unit testing (testing a class in isolation of the others). It assists in adhering to unit-testing best practices by randomizing test execution order out of the box. Therefore special care has to be taken if TestNG should be used to execute integration-tests that require to run isolated by class in a specific test-order.

Preferably there would be a section on how to best implement integration-tests with TestNG or a pointer to an alternative library which is better suited for that scenario.

Thank you @cbeust for your amazing work on TestNG!

**Comment by juherr** [comment]
@michabu I don't understand what is the problem and what you expect. Could you be more precise? Maybe you should open a new issue if you have a specific sample to share.

**Comment by cbeust** [comment]
Hi Michael, and sorry you're running into problems.

I'd like to take issue with this statement:

> Prominently and explicitly state in the docs that TestNG is not to be
used with stateful integration-tests or introduce a section on how to use
it in that case

I created TestNG purposefully to cover integration and functional tests,
with unit tests being a possibility as well as a simple consequence of
TestNG being a testing framework.

TestNG offers a variety of options if you have strict ordering
requirements, with dependencies (method and group dependencies) being the
most popular one, but priorities being an option as well.

Interleaving is tricky to achieve reliably with all these sequencing
options being available (remember that TestNG allows you to group
executions by suites, classes, groups, classes and instances) so if the
ordering that TestNG comes up with doesn't suit you, you should simply
impose the ordering you desire with dependsOnGroups..


-- 
Cédric

**Comment by michabu** [comment]
Thank you for your quick reply @juherr, @cbeust 

Maybe I'm missing something. Let me explain my requirements to you and figure that out. And let me also offer to help improving the documentation if it is only a matter of understanding.

On the one hand I have unit tests. Those are independent of each other. Random (and parallel) execution and interleaving of Tests helps a) speed up test execution b) discover accidental dependencies between test. -> This works really well for me.

On the other hand I have integration tests. Those tests can get quite lengthily which is why I divide them into a sequence of steps that are dependent on each other (using `dependsOnMethods`).
The division of   one integration test into several methods caters for readability, maintainability and easy analysis of what went wrong in such a scenario.
For some of the integration-tests parts of the system are run in different threads which arises the need to shutdown those parts before executing the next test. I currently achieve that by having a `shutdownTest`-Method which `dependsOnMethods="", alwaysRun=true`.
Database-creation is expensive therefore all integration-tests share one freshly created database. I achieve that by having all of the integration tests in the same group. E.g. `groups="integration-tests"`.
One integration-test resides in one class. That allows to clearly state the purpose of the test (the name of the class) and the single steps involved (the name of the methods the test is composed of).

What I'm struggling with is to get the methods in one class to execute in the defined order without them being interleaved by methods of other test-classes.

Quoting the example from @cbeust on 8 Oct 2011 <- This is the behaviour I observer on our build server.

> Running: 1.1
> Running: 2.3
> Running: 2.2
> Running: 2.4
> Running: 2.1
> Running: 1.2
> Running: 1.3

The behaviour needed for the integration-tests to run in a meaningful manner is:

> Running: 1.1
> Running: 1.2
> Running: 1.3
> Running: 2.1
> Running: 2.2
> Running: 2.3
> Running: 2.4

The integration-tests in this case would look like

```java
public class c1 {
    @Test(
        groups = "integration-tests"
    )
    public void m1 {
    }

    @Test(
        groups = "integration-tests",
        dependsOnMethods = "m1"
    )
    public void m2 {
    }

    @Test(
        groups = "integration-tests",
        dependsOnMethods = "m2"
    )
    public void m3 {
    }

    @Test(
        groups = "integration-tests",
        alwaysRun = true,
        dependsOnMethods = "m3"
    )
    public void shutdownTest {
    }
}

public class c2 {
    @Test(
        groups = "integration-tests"
    )
    public void m1 {
    }

    @Test(
        groups = "integration-tests",
        dependsOnMethods = "m1"
    )
    public void m2 {
    }

    @Test(
        groups = "integration-tests",
        dependsOnMethods = "m2"
    )
    public void m3 {
    }

    @Test(
        groups = "integration-tests",
        dependsOnMethods = "m3"
    )
    public void m4 {
    }

    @Test(
        groups = "integration-tests",
        alwaysRun = true,
        dependsOnMethods = "m4"
    )
    public void shutdownTest {
    }
}
```

For completeness there is also a class that creates the database for all the tests:
```java
public class IntegrationTestProvisioning {
    private static final Logger log = LoggerFactory
        .getLogger(IntegrationTestProvisioning.class.getName());

    @BeforeGroups(
        groups = {
            "integration-tests"
        }
    )
    public void beforeGroup() {
        log.info("Creating database for group integration-tests");
    }
}
```

Let me emphasis that there is no dependency between the integration-test classes. That is all test-classes should be run. Whether a test-class fails or not should not make a difference in running other test-classes.

I'm looking forward to your answers @cbeust, @juherr. Thank you for taking the time.

**Comment by juherr** [comment]
What is the parallel value you use for integration?
`group-by-instance` should make the job, but it may have some issue depending on your parallel value.

**Comment by michael-buerkle** [comment]
Now, this really took me a long time to get back to. Thank you nevertheless very much for your help.

I think my main problem was to understand how to apply `group-by-instance` which is frankly quite embarrassing. But here you go:

This is how my setup looks like now in order to get the expected behaviour:

`build.gradle`:
```groovy
test {
    useTestNG() {
        suites 'src/test/resources/testng.xml'
    }
}

dependencies {
    testCompile group: 'org.testng', name: 'testng', version: '6.14.3'
}
```
`src/test/resources/testng.xml`:
```xml
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
<suite name="all-tests" verbose="1">
    <test name="unit-tests" parallel="methods" preserve-order="false">
        <groups>
            <run>
                <include name="unit-test"/>
            </run>
        </groups>

        <packages>
            <package name="com.my.test.package.*"/>
        </packages>
    </test>

    <test name="integration-tests" parallel="false" group-by-instances="true">
        <groups>
            <run>
                <include name="integration-test"/>
            </run>
        </groups>

        <packages>
            <package name="com.my.test.package.*"/>
        </packages>
    </test>
</suite>
```

I hope this might help someone else who happens to stumble over the same issue.
Thank you for making testNG work!

**Comment by michabu** [comment]
Now, this really took me a long time to get back to. Thank you nevertheless very much for your help.

I think my main problem was to understand how to apply `group-by-instance` which is frankly quite embarrassing. But here you go:

This is how my setup looks like now in order to get the expected behaviour:

`build.gradle`:
```groovy
test {
    useTestNG() {
        suites 'src/test/resources/testng.xml'
    }
}

dependencies {
    testCompile group: 'org.testng', name: 'testng', version: '6.14.3'
}
```
`src/test/resources/testng.xml`:
```xml
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
<suite name="all-tests" verbose="1">
    <test name="unit-tests" parallel="methods" preserve-order="false">
        <groups>
            <run>
                <include name="unit-test"/>
            </run>
        </groups>

        <packages>
            <package name="com.my.test.package.*"/>
        </packages>
    </test>

    <test name="integration-tests" parallel="false" group-by-instances="true">
        <groups>
            <run>
                <include name="integration-test"/>
            </run>
        </groups>

        <packages>
            <package name="com.my.test.package.*"/>
        </packages>
    </test>
</suite>
```

I hope this might help someone else who happens to stumble over the same issue.
Thank you for making testNG work!
